### PR TITLE
Backwards compatibility for orchestrationClient

### PR DESCRIPTION
**Pull request by davidmrdavid**
This PR addresses: https://github.com/Azure/azure-functions-durable-extension/issues/1290

There changes are few and simple (_except I apparently pushed other people's commits, more on this later_):

1. `DurableClientAttribute` is no longer a `sealed` class, so we can subclass it.

2. The declaration of a new dummy subclass of `DurableClientAttribute` named `OrchestrationClientAttribute`

3. A new binding rule in `DurableTaskExtension.cs` that explicitly accepts `OrchestrationClientAttribute`. This solution is less than ideal, because there's a handful of copy-pasted lines from the rule accepting `DurableClientAttribute`. Ideally, we would have a single line allowing us to select between the two attributes. If anyone knows how to do this, *please let me know*

So, while this works, I'm still looking for a solution with less repetition. I'm open to input!

Finally, in  https://github.com/Azure/azure-functions-durable-extension/issues/1290 we discuss the possibility of raising warnings on the logs of `func host start` for JS users using `orchestrationClient` instead of `durableClient`. This PR assumes that this is change we would implement exclusively on the JS side of things. If this is not the case, please let me know.

### CODE REVIEW

**Comment by davidmrdavid** [comment]
Hmm, I see a few too many commits being pushed. Perhaps I pulled from the wrong source? Sorry about that, I'll look into fixing that next!

**Comment by cgillum** [comment]
Yeah, that's definitely weird. I typically see it if by branches aren't fully up-to-date. I don't yet really understand what causes it though.

One thing I would suggest is to exclude your changes under the `docs` directory. This is actually all the auto-generated documentation that existed before we migrated to docs.microsoft.com. We haven't gotten around to deleting this content from our repo. 😅

**Comment by cgillum** [comment]
Also, it looks like the AppVeyor build is failing due to the usage of an obsolete API. 

>   DurableTaskExtension.cs(259,60): error CS0618: 'OrchestrationClientAttribute' is obsolete: '{OrchestrationClientAttribute is obsolete. Use DurableClientAttribute instead.' [C:\projects\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\WebJobs.Extensions.DurableTask.csproj]

We should suppress that warning in the source code since we're doing this intentionally. I think Visual Studio should have some tool tips to help suppress warnings in source code.

**Comment by davidmrdavid** [comment]
After *alot* of work trying to fix my git state, the diff appears right now. Sorry for the clutter folks, I'll be more careful next time. My mistake was to merge accidentally master and to repeatedly fail to perform a `git diff` from local to remote.

**Comment by cgillum** [comment]
Glad to see everything is working! :)

However, I'm still a little concerned about the Git history represented in this PR. If you're certain that your local branch is good, would you mind locally squashing your changes and doing a force push to make sure that GitHub also thinks things are looking clean?

I imaging the following commands could work for you assuming your branch is in a sane state. They will do the following:

1. Rebase your commits on top of the latest from `dev` to ensure maximum cleanliness
2.  Squash your changes into a single commit
3. Overwrite the history of this PR to match the *new* history of your local branch

```bash
git checkout dev
git pull
git checkout djusto/backwards_compatibility
git rebase dev

git reset --soft 4b81619ddd31043f63a1e351e33367c8de65f368
git commit -m "Backwards compatibility for orchestrationClient"

git push --force
```

**Comment by davidmrdavid** [comment]
@cgillum , thanks for the helpful commands and their explanation. I spent a little bit of time today also reading up on how to get out of these bad git states. Anyhow, it should be looking better now :)

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs [hunk 1]
```diff
@@ -252,6 +252,22 @@ void IExtensionConfigProvider.Initialize(ExtensionConfigContext context)
             rule.BindToInput<IDurableEntityClient>(this.GetClient);
             rule.BindToInput<IDurableClient>(this.GetClient);
 
+            // We add a binding rule to support the now deprecated `orchestrationClient` binding
+            // A cleaner solution would be to have the prior rule have an OR-style selector between
+            // DurableClientAttribute and OrchestrationClientAttribute, but it's unclear if that's
+            // possible (for now).
+#pragma warning disable CS0618 // Type or member is obsolete
+            var backwardsCompRule = context.AddBindingRule<OrchestrationClientAttribute>()
```
Comment (amdeel) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: Why is this needed? Shouldn't this work without these changes with inheritance?

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs [hunk 2]
```diff
@@ -252,6 +252,22 @@ void IExtensionConfigProvider.Initialize(ExtensionConfigContext context)
...
+#pragma warning disable CS0618 // Type or member is obsolete
+            var backwardsCompRule = context.AddBindingRule<OrchestrationClientAttribute>()
+#pragma warning restore CS0618 // Type or member is obsolete
+                .AddConverter<string, StartOrchestrationArgs>(bindings.StringToStartOrchestrationArgs)
+                .AddConverter<JObject, StartOrchestrationArgs>(bindings.JObjectToStartOrchestrationArgs)
+                .AddConverter<IDurableClient, string>(bindings.DurableOrchestrationClientToString);
+
+            backwardsCompRule.BindToCollector<StartOrchestrationArgs>(bindings.CreateAsyncCollector);
+            backwardsCompRule.BindToInput<IDurableOrchestrationClient>(this.GetClient);
+            backwardsCompRule.BindToInput<IDurableEntityClient>(this.GetClient);
```
[the remark this answers is not in the document]
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: Note that if you go the route where you get rid of `backwardsCompRule`, you can disregard this feedback.

[the remark this answers is not in the document]
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: We already discourage users from using `[OrchestrationClient]` by putting the `[Obsolete]` attribute on it.  But the whole reason for `backCompRule` is that inheritance by itself is not sufficient, per David's investigation last week.

[the remark this answers is not in the document]
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: Good to know inheritance didn't work. I must either be misremembering the behavior of binding rules, or they have changed since I last looked at the code.

In that case, while I still don't see a good reason to include support for the backwards compatible binding with a V2 only feature, it probably doesn't hurt to include the support, so this nit can be disregarded.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs (hunk 1 above)
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: As mentioned on another issue, it sounds like you already tried removing this, but it didn't work.

Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: @amdeel @ConnorMcMahon 
Yup! This code is only here because inheritance didn't work by itself. 
It appears that inheritance is enough to _register_ the binding, so a `func host start` runs successfully. However, invoking the function will trigger a runtime exception, so inheritance-alone only seems to take us half of the way there.

I don't think this is the expected behaviour, since it surprises all of us. I would love to follow-up with the WebJobs folks to tell them about our usecase

**Comment by davidmrdavid** [comment]
My latest commit(s) adds _quite simple_ unit tests as a sanity check that my changes work. They are slight variants on already existing tests for the `DurableClientAttribute`.

I also manually tested the usage of `orchestrationClient` as a binding in JS and C#, and a handful of samples seem to continue working. All in all, I think this is ready.

I'll wait for another 'lgtm' and them I'll squash-and-merge. Thanks folks!

File: test/Common/DurableTaskHostExtensions.cs [hunk 3]
```diff
@@ -96,5 +96,47 @@ internal static class DurableTaskHostExtensions
             IDurableOrchestrationClient client = clientRef[0];
             return client;
         }
+
+        /// <summary>
+        /// Helper function for the IDurableEntityClientBindingBackComp test. Gets an IDurableEntityClient.
+        /// </summary>
+        public static async Task<IDurableEntityClient> GetEntityClientBindingBackCompTest(
```
Comment (ConnorMcMahon) [inline test/Common/DurableTaskHostExtensions.cs:103]: nit: I realize you just copied the approach from previous tests, but since this logic is only used for one test, I think I prefer having it actually appear in the test instead of in a helper method. Otherwise just looking at the test doesn't give us a clear idea of what is happening.

Comment (davidmrdavid) [reply test/Common/DurableTaskHostExtensions.cs:103]: Sounds good, will fix and merge!
